File: CSES/Advertisement/soln.cpp
/*
 * written by Yaqi Zhang (zhang623@wisc.edu)
 * University of Wisconsin-Madison
 * April 2019
 */
#include <bits/stdc++.h>

using namespace std;

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(0);
    int n;
    cin >> n;
    vector<int> hs(n + 2);
    hs[0] = -1;
    for(int i = 0; i < n; ++i) cin >> hs[i + 1];
    hs[n + 1] = -1;
    stack<int> stk;
    long long mx_area = 0LL;
    for(int i = 0; i < n + 2; ++i) {
        while (!stk.empty() && hs[i] < hs[stk.top()]) {
            int h = hs[stk.top()];
            stk.pop();
            int w = i - stk.top() - 1;
            mx_area = max(mx_area, (long long) h * w);
        }
        stk.push(i);
    }
    cout << mx_area << endl;
    return 0;
}

File: CSES/Advertisement/soln.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
##################################
# University of Wisconsin-Madison
# Author: Yaqi Zhang
##################################
# This module contains ...
##################################

# standard library
import sys

def main():
    n = int(input())
    hs = list(map(int, input().split()))
    stack = [-1]
    hs.append(0)
    mx = 0
    for i, h in enumerate(hs):
        while stack and h < hs[stack[-1]]:
            height = hs[stack.pop()]
            width = i - stack[-1] - 1
            mx = max(mx, height * width)
        stack.append(i)
    print(mx)


if __name__ == "__main__":
    main()

